### PR TITLE
WIP Custom headers in Java DSL

### PR DESCRIPTION
**Pull request by johanandren**

### CODE REVIEW

File: akka-http-tests/src/test/java/akka/http/javadsl/server/directives/HeaderDirectivesTest.java [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+    route
+      .run(HttpRequest.create().addHeader(new MyHeader("such-value")))
+      .assertStatusCode(StatusCodes.OK)
+      .assertEntity("such-value");
+
+    route
+      .run(HttpRequest.create())
+      .assertStatusCode(StatusCodes.BAD_REQUEST);
+  }
```
Comment (johanandren) [inline akka-http-tests/src/test/java/akka/http/javadsl/server/directives/HeaderDirectivesTest.java]: this test case passes just fine

Comment (johanandren) [inline akka-http-tests/src/test/java/akka/http/javadsl/server/directives/HeaderDirectivesTest.java:183]: I meant, _this_ test case passes just fine.

File: akka-http-tests/src/test/java/akka/http/javadsl/server/directives/HeaderDirectivesTest.java [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+      .assertStatusCode(StatusCodes.OK)
+      .assertEntity("false");
+  }
+
+
+  public static class MyHeader extends ModeledCustomHeader {
+
+    private final String value;
+
+    public MyHeader(String value) {
```
Comment (ktoso) [inline akka-http-tests/src/test/java/akka/http/javadsl/server/directives/HeaderDirectivesTest.java]: This is accidental though I think.
Make the MyHeader "weird", as in: `public MyHeader(Long shoeSize, String name) {` and hit it with a request like: `X-My-Header: 42-link`

File: akka-http-tests/src/test/java/akka/http/javadsl/server/directives/HeaderDirectivesTest.java [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+
+    route
+      .run(HttpRequest.create().addHeader(new ShoeHeader("boots", 42L)))
+      .assertStatusCode(StatusCodes.OK)
+      .assertEntity("boots-42");
+
+    route
+      .run(HttpRequest.create())
+      .assertStatusCode(StatusCodes.BAD_REQUEST);
+  }
```
Comment (johanandren) [inline akka-http-tests/src/test/java/akka/http/javadsl/server/directives/HeaderDirectivesTest.java:183]: this test still passes just fine

File: akka-http-tests/src/test/java/akka/http/javadsl/server/directives/HeaderDirectivesTest.java (hunk 1 above)
Comment (johanandren) [inline akka-http-tests/src/test/java/akka/http/javadsl/server/directives/HeaderDirectivesTest.java:183]: Argh, github-diff-view why you move my comments? 😡

File: akka-http-tests/src/test/java/akka/http/javadsl/server/directives/HeaderDirectivesTest.java [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+
+  }
+
+  @Test
+  public void testValueByTypeHandlesCustomHeaders() {
+    TestRoute route = testRoute(headerValueByType(MyHeader.class,
+      (MyHeader m) -> complete(m.value())));
+
+    route
+      .run(HttpRequest.create().addHeader(new MyHeader("such-value")))
```
Comment (ktoso) [inline akka-http-tests/src/test/java/akka/http/javadsl/server/directives/HeaderDirectivesTest.java]: This is lucky to work. Pass in a raw header and see it fail.

File: akka-http/src/main/scala/akka/http/scaladsl/server/directives/HeaderDirectives.scala [hunk 5]
```diff
@@ -177,6 +177,6 @@ trait LowPriorityHeaderMagnetImplicits {
     new HeaderMagnet[T] {
       val classTag: ClassTag[T] = tag
       val runtimeClass: Class[T] = tag.runtimeClass.asInstanceOf[Class[T]]
-      val extractPF: PartialFunction[Any, T] = { case x: T if runtimeClass.isAssignableFrom(x.getClass) ⇒ x }
+      val extractPF: PartialFunction[Any, T] = { case x if runtimeClass.isAssignableFrom(x.getClass) ⇒ x.asInstanceOf[T] }
     }
 }
```
Comment (johanandren) [inline akka-http/src/main/scala/akka/http/scaladsl/server/directives/HeaderDirectives.scala:182]: change we talked about a couple of days ago, thought I'd just boyscout them.

File: akka-http/src/main/scala/akka/http/javadsl/server/directives/HeaderDirectives.scala [hunk 6]
```diff
@@ -61,10 +66,31 @@ abstract class HeaderDirectives extends FutureDirectives {
...
+      val magnet: HeaderMagnet[_] =
+        refl.getObjectFor[ModeledCustomHeaderCompanion[_]](t.getName) match {
+          case Success(companion) =>
+            new HeaderMagnet[T] {
+              override def classTag = ClassTag(t)
+              override def runtimeClass = t
+              override def extractPF = {
+                case h if h.is(companion.lowercaseName) => companion.apply(h.toString).asInstanceOf[T]
+              }
+            }
```
Comment (johanandren) [inline akka-http/src/main/scala/akka/http/javadsl/server/directives/HeaderDirectives.scala:78]: Couldn't figure out how to lie enough convincingly to scalac about being sure the type parameter `T <: ModeledCustomHeader[T]` so I couldn't use `HeaderMagnet.fromClassForModeledCustomHeader` here

File: akka-http/src/main/scala/akka/http/scaladsl/server/directives/HeaderDirectives.scala (hunk 5 above)
Comment (ktoso) [inline akka-http/src/main/scala/akka/http/scaladsl/server/directives/HeaderDirectives.scala:182]: yes, great 👍

File: akka-http/src/main/scala/akka/http/javadsl/server/directives/HeaderDirectives.scala [hunk 7]
```diff
@@ -55,17 +60,35 @@ abstract class HeaderDirectives extends FutureDirectives {
...
    */
   def headerValueByType[T <: HttpHeader](t: Class[T], inner: jf.Function[T, Route]) = RouteAdapter {
-    // TODO custom headers don't work yet
-    // TODO needs instance of check if it's a modeled header and then magically locate companion
-    D.headerValueByType(HeaderMagnet.fromClassNormalJavaHeader(t)) { value =>
+
+    def magnetForModeledCustomHeader(clazz: Class[T]): HeaderMagnet[T] = {
+      // figure out the modeled header companion and use that to parse the header
+      val refl = new ReflectiveDynamicAccess(getClass.getClassLoader)
+      refl.getObjectFor[ModeledCustomHeaderCompanion[_]](t.getName) match {
```
Comment (ktoso) [inline akka-http/src/main/scala/akka/http/javadsl/server/directives/HeaderDirectives.scala:71]: exactly, thanks

**Comment by ktoso** [comment]
Well done 👍 
Semi-resolves https://github.com/akka/akka/issues/20358

**Pull request merged**